Add dev container

ShimShtein opened the pull request.

pvoborni commented:
Hi thanks, the PR might get better description of what is the expected functionality. My assumption/expectation is to execute `host-metering once` successfully. 

But if I run this (click on `Run once [mocked]` in `Run and Debug` section) I get the following output:

```
Updating config from config file...
Failed to process file: no config file at /etc/host-metering.conf
Updating config from environment variables...

Config:
  WriteUrl: http://localhost:9090/api/v1/write
  WriteIntervalSec: 600
  HostCertPath: $REDACTED/host-metering/mocks/consumer/cert.pem
  HostCertKeyPath: $REDACTED/host-metering/mocks/consumer/key.pem
  CollectIntervalSec: 0
  LabelRefreshIntervalSec: 86400
  WriteRetryAttempts: 8
  WriteRetryMinIntSec: 1
  WriteRetryMaxIntSec: 10
  WriteTimeoutSec: 60
  MetricsMaxAgeSec: 5400
  MetricsWALPath: /var/run/host-metering/metrics
  LogLevel: INFO
  LogPath: 
Cert Watcher initialization failed: no such file or directory
mkdir /var/run/host-metering: permission denied
Failed to create daemon: mkdir /var/run/host-metering: permission denied
```

which indicates, that

1. `MetricsWALPath` config value should be adjusted to something where the app can write (not the default `/var/run/host-metering` - e.g. via `HOST_METERING_METRICS_WAL_PATH` env var.

But I expect that:

1. "notification" will fail on not having `http://localhost:9090/api/v1/write` accessible, i.e. it would need to run a Prometheus server in a container. 
2.  I did not do anything with subscription manager. So it might fail on that part. It would be great if the subscription could be completely mocked without a having it on the host system. 

Also please rebase the PR.

pvoborni commented on mocks/create-cert.sh:
```diff
@@ -0,0 +1,20 @@
+
```
The repo already contains:

- `create-cert.sh`
- `host-metering.conf`
- `premetheus.yml`
-  `test-cert.cnf`

files in `hack` directory: https://github.com/RedHatInsights/host-metering/tree/main/hack 

Also  mocked `subscription-manager` is in tests dir: https://github.com/RedHatInsights/host-metering/blob/main/test/bin/subscription-manager 

Is there a reason to re-add them and not delete the old?  Especially when the PR contains changes to stop using the old.

pvoborni commented on .devcontainer/host-metering.conf:
```diff
@@ -0,0 +1,17 @@
+# E.g. could be used as:
+# ./host-metering -config hack/host-metering.conf daemon
+
+[host-metering]
```
This file is based on older configuration which doesn't work with the latest changes. The latest one is updated. 

It produces issue:
```
Invalid configuration: WriteInterval must be bigger than WriteRetryAttempts * ( WriteRetryMaxInt + WriteTimeout )
```

It doesn't stop on this error with e.g.:

```diff
diff --git a/.devcontainer/host-metering.conf b/.devcontainer/host-metering.conf
index 7035589..b5bec40 100644
--- a/.devcontainer/host-metering.conf
+++ b/.devcontainer/host-metering.conf
@@ -11,7 +11,8 @@ host_cert_path=./mocks/consumer/cert.pem
 host_cert_key_path=./mocks/consumer/key.pem
 collect_interval_sec=0
 label_refresh_interval_sec=10
-write_retry_attempts=4
-write_retry_interval_sec=2
+write_timeout_sec=1
+write_retry_attempts=1
+write_retry_max_int_sec=2
```

pvoborni commented on .devcontainer/host-metering.conf:
```diff
@@ -0,0 +1,17 @@
+# ./host-metering -config hack/host-metering.conf daemon
+
+[host-metering]
+write_url=http://prometheus:9090/api/v1/write
```
The url doesn't work when I run it with `make test-daemon`. It starts working when changed back to `write_url=http://localhost:9090/api/v1/write`  (tried in env which had all podman pods removed/cleaned).

pvoborni replied:
To improve part of the performance degradation. The `host-metering.conf` could use `localhost` and only in the "compose" envs override it with `HOST_METERING_WRITE_URL`, e.g by setting this env var in `.vscode/launch.json` `env` var part or in the `docker-compose.yml` in `environment` for `host-metering` service.  This would also mean to not use the compose for  `make test-daemon`.

pvoborni commented on Makefile:
```diff
@@ -53,51 +54,48 @@ build-selinux:
 
 .PHONY: prometheus
-prometheus: container-pod
+prometheus: podman-containers
```
I don't like that using `podman-containers` and thus `podman-compose` target causes that `test-daemon` start is delayed by `2-3s` (the build of compose) instead of starting almost instantly.  This is annoying in development.  And would be better if `make test-daemon` target would not be involved in the compose at all.

pvoborni commented on mocks/mock_from_host.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
```
Where is this file used?

pvoborni commented on README.md:
```diff
@@ -86,3 +86,28 @@ Or visit the Prometheus Web UI at http://localhost:9090/graph?g0.expr=system_cpu
+`mocked_run.sh` is a wrapper of `go run main.go` to execute the command in mocked context.
+
+### Preparing the mocks
+Inside the `mocks` folder run `./prepare_mock.sh` to generate outputs that will be used as mocks.
```
There is no `./prepare_mock.sh`.  Was `mock_from_host.sh` meant? And would it work?

pvoborni commented on README.md:
```diff
@@ -86,3 +86,28 @@ Or visit the Prometheus Web UI at http://localhost:9090/graph?g0.expr=system_cpu
+There is an option to run make commands inside the `docker-compose` generated environment. Just prefix a make command you would like to run with `podman-`. e.g. to run `make test` in a container, use `make podman-test`.
+
+## Mocking subscription-manager commands
+`mocked_run.sh` is a wrapper of `go run main.go` to execute the command in mocked context.
```
What's the purpose of `mocked_run.sh` ?  It does almost the same thing as `make test-daemon` but it cannot:

- start prometheus if it is not running
- adjust configuration bit easier by modifying `host-metering.conf` (though it's possible to adust it by adding more env vars)

pvoborni replied:
The previous comment still applies - this section doesn't describe why to use it.

pvoborni reviewed (commented):
I tried to run it in vscode dev container in remote vs. The README looks good  - it helped to setup it. 

But when I run it, it ends with:
```
Metrics collected
Initiating notification request...
Sending 21 sample(s)...
Error calling PrometheusRemoteWrite: PrometheusRemoteWrite: Post "http://prometheus:9090/api/v1/write": dial tcp: lookup prometheus on REDACTED_IP: no such host
Process 5563 has exited with status 0
Detaching
dlv dap (5483) exited with code: 0
```

Podman shows:
```
podman ps -a
CONTAINER ID  IMAGE                                                    COMMAND               CREATED        STATUS        PORTS                   NAMES
686df1121895  localhost/hostmeteringdevcontainer_host-metering:latest  /bin/sh -c while ...  4 minutes ago  Up 4 minutes                          hostmeteringdevcontainer_host-metering_1
b442fa1ea972  quay.io/prometheus/prometheus:latest                     --config.file=/et...  4 minutes ago  Created       0.0.0.0:9090->9090/tcp  hostmeteringdevcontainer_prometheus_1
```

Does it work on your side?

Also running go test in the vscode UI did not work almost at all for me.  It works well in non dev-container (except the tests which relies on fake subman in path). 

Technicality: the "feat" prefix is not suitable for these commits as they are not adding any feature to the software being build. They change how we develop it. So e.g.  "chore", "ci" would be better.  https://github.com/RedHatInsights/host-metering/blob/main/CONTRIBUTING.md#type

ShimShtein commented:
Addressed all the comments, ready for another round. 
I have tried testing with vscode and with `make podman-test` and it seems to work as expected, even the tests that depend on mocked sub-man are working fine for me.

There is a slight performance issue since make reads all the `ifeq` statements before deciding which rule to run, but I suggest leaving it to another commit.

pvoborni commented:
I did some testing and the performance degradation is not "slight" it is huge, e.g. 7-30 times slower. Mostly caused by the `getent` call and timeout when it is not found. This degrading my developer experience a lot and thus the patch should not go in unless this is resolved to times similar to now. 

With current `main`:

Timings (on my non-so-great virtual machine):
- `make prometheus`: 0.384s
- `make test`: 1.3s
- `make test-daemon` (clean env): 1.3s - 3.18s (depends if prometheus starts before first request)
- `make test-daemon` (second run): 0.23s
- `make clean-all`: 0.02s - 0.2s

Note: `make test-daemon` is measure when command is changed from `daemon` to  `once`.


With this patch:
- `make prometheus`: 9.81s
- `make test`: 9.3s
- `make test-daemon` (clean env): 10.43s
- `make test-daemon` (second run): 10.68s
- `make clean-all`: 8.48s 


From my PoV:
- I don't see value in using docker/podman-compose it just slows down current calls which only need a container with podman running
- I don't have anything against adding the VSCode devel containers, but they should not negatively influence the current workflows

ShimShtein commented:
@pvoborni I was able to speed up the build. Now make commands run pretty much instantly

pvoborni reviewed (approved):
@ShimShtein  thanks for the improvements, this is much better. The PR LGTM, though I did not re-test the devel container again as it doesn't work the best on my setup, but I believe it works for you. 

Some nits, I feel obligated to mention, but this PR can be merged anyway:
- a lot of the files are missing empty line at the end of file - which is a good convention to keep
-  I'd keep the comment with usage in `mocks/subscription-manager` as it documents the "CLOUD_PROVIDER" env var.
